src/DistributedArrays.jl

```diff
@@ -880,13 +880,37 @@ for f in (:+, :-, :div, :mod, :rem, :&, :|, :$)
             B = samedist(A, B)
             map_localparts($f, A, B)
         end
+        function ($f){T}(A::DArray{T}, B::Array{T})
+            size(A) == size(B) || throw(DimensionMismatch("first array has size $(size(A)) but second array has $(size(B))"))
+            return DArray(A) do I
+                ($f)(localpart(A), B[I...])
```

Comment:
The easy thing to try out would be 
```
ref = RemoteChannel()
put!(ref, B)
b_pid = myid()
return DArray(A) do I
     ($f)(localpart(A), remotecall_fetch((r,idxs)->fetch(r)[idxs...], b_pid, ref, I))
end
```

This would call back onto the process hosting B (typically 1) and retrieve only the part of B required on the local worker.

Further optimization where only the required part of B is sent to each node as part of the init function itself can be achieved by writing a custom serializer and the `construct_darray` function.

A type to capture state:
```
type SmartArraySend
   A::Array
   idxs::Nullable{Array}
   pids::Nullable{Array}
end 

Base.serialize(S::SerializationState, sa::SmartArraySend)
  pid = Base.worker_id_from_socket(s.io)
  I = sa.idxs[findfirst(sa.pids)]
  Serializer.serialize_type(S, typeof(sa))
  serialize(S, sa.A[I...]) 
end

Base.deserialize{T<:SmartArraySend}(S::SerializationState, t::Type{T})
    A_part = deserialize(S)
    return SmartArraySend(A_part, Nullable{Array}(), Nullable{Array}())
end
```


And the modified constructor:

```
sa = SmartArraySend(B, A.idxs, A.pids)
return DArray(A) do I
     ($f)(localpart(A), getfield(sa,:A))
end
```